File: TypeScript/app/updateStrategies/DefaultStrategy.ts
import Item from "./../models/Item";
import IUpdateStrategy from "./IUpdateStrategy";


export default class DefaultStrategy implements IUpdateStrategy {    
    constructor(private qualityFactor : number) {}

    updateItem(item: Item) : Item {
        let newQuality = Math.max(0, item.quality - this.qualityFactor);
        let newSellIn = item.sellIn - 1;
        return new Item(item.name, newSellIn, newQuality);
    }
}

File: Java/src/main/java/com/gildedrose/DefaultStrategy.java
package com.gildedrose;

public enum DefaultStrategy implements QualityUpdateStrategy {
    INSTANCE;

    @Override
    public void applyTo(Item item) {
        item.quality = Math.max(0, item.sellIn > 0 ? (item.quality-1) : (item.quality-2));
        item.sellIn = item.sellIn - 1;
    }
}
